go/service/chat_local.go

```diff
@@ -89,58 +175,149 @@ func (h *chatLocalHandler) CompleteAndCanonicalizeTlfName(ctx context.Context, t
-func (h *chatLocalHandler) ResolveConversationLocal(ctx context.Context, arg keybase1.ConversationInfoLocal) (ids []chat1.ConversationID, err error) {
-	res, err := h.boxer.tlf.CryptKeys(ctx, arg.TlfName)
-	if err != nil {
-		return ids, err
+// keybase.chatLocal.GetOrCreateTextConversationLocal protocol.
+func (h *chatLocalHandler) ResolveConversationLocal(ctx context.Context, arg keybase1.ConversationInfoLocal) (conversations []keybase1.ConversationInfoLocal, err error) {
+	if arg.Id != 0 {
+		info, _, err := h.getConversationInfoByID(ctx, arg.Id)
+		if err != nil {
+			return nil, err
+		}
+		return []keybase1.ConversationInfoLocal{info}, nil
 	}
-	tlfIDb, err := hex.DecodeString(string(res.TlfID))
-	if err != nil {
-		return ids, err
+	return h.searchForConversations(ctx, arg)
+}
+
+// searchForConversations searches for conversations using tlfName, topicName,
+// and topicType fields in criteria, and returns all matching conversations.
+// Conversation IDs are populated in returned conversations.
+func (h *chatLocalHandler) searchForConversations(ctx context.Context, criteria keybase1.ConversationInfoLocal) (conversations []keybase1.ConversationInfoLocal, err error) {
+	appendMaybe := func(info keybase1.ConversationInfoLocal) {
+		if len(criteria.TopicName) > 0 && criteria.TopicName != info.TopicName {
+			return
+		}
+		if criteria.TopicType != chat1.TopicType_NONE && criteria.TopicType != info.TopicType {
+			return
+		}
+		conversations = append(conversations, info)
 	}
-	tlfID := chat1.TLFID(tlfIDb)
 
-	ipagination := &chat1.Pagination{Num: 20}
-getinbox:
-	for i := 0; i < 10000; /* in case we have a server bug */ i++ {
-		iview, err := h.GetInboxLocal(ctx, ipagination)
+	if len(criteria.TlfName) > 0 { // fast case
+		// TODO: do some caching in boxer so we don't end up calling this RPC
+		// unnecessarily too often
+		resp, err := h.boxer.tlf.CryptKeys(ctx, criteria.TlfName)
 		if err != nil {
-			return ids, err
+			return nil, err
+		}
+		tlfIDb, err := hex.DecodeString(string(resp.TlfID))
+		if err != nil {
+			return nil, err
+		}
+		conversationsRemote, err := h.remoteClient().GetInboxByTLFIDRemote(ctx, tlfIDb)
+		if err != nil {
+			return nil, err
 		}
-		for _, conv := range iview.Conversations {
-			if conv.Metadata.IdTriple.Tlfid.Eq(tlfID) {
-				// TODO: check topic name and topic ID here when we support multiple
-				// topics per TLF
-				ids = append(ids, conv.Metadata.ConversationID)
+		for _, cr := range conversationsRemote {
+			info, _, err := h.getConversationInfo(ctx, cr)
+			if err != nil {
+				return nil, err
+			}
+			if info.TlfName != criteria.TlfName {
+				// check again using signed information to make sure it's the correct
+				// conversation
+				return nil, libkb.UnexpectedChatDataFromServer{Msg: fmt.Sprintf("Unexpected data is returned from server. We asked for %v, but got conversation for %v. TODO: handle tlfName changes properly for SBS case", criteria.TlfName, info.TlfName)}
```

Comment:
I was thinking about this case some more. I think the way we will end up having to do this is everytime there is an SBS resolution, MD server needs to let Gregor know about it too (like it does with kbfstlfd). When it does, Gregor needs to generate a TLFNAME message for all conversations on the TLF ID. Then, in `getConversationInfo` we return the complete list all of these messages and return a list of TLF names to check the criteria against.